### PR TITLE
Basic support for JWK and headers

### PR DESCRIPTION
**Pull request by csstaub**
This adds basic support for the JWK format, and allows for reading (some, not all) header values from parsed messages. Please take a look.

R: @dgalling @sul3n3t
CC: @rapropos @bifurcation

### CODE REVIEW

**Comment by csstaub** [comment]
(This should address issues #2, #3 and #7)

File: asymmetric.go [hunk 1]
```diff
@@ -219,7 +226,7 @@ func (ctx rsaDecrypterSigner) signPayload(payload []byte, alg SignatureAlgorithm
 	case RS512, PS512:
 		hash = crypto.SHA512
 	default:
-		return signatureInfo{}, ErrUnsupportedAlgorithm
+		return Signature{}, ErrUnsupportedAlgorithm
```
Comment (dgalling) [inline asymmetric.go]: Should this return `nil, ErrUnsupportedAlgorithm`?

Comment (csstaub) [inline asymmetric.go]: Can't return `nil` here, return value is not a pointer.

Comment (dgalling) [inline asymmetric.go]: Ah, duh.

File: jwe.go [hunk 2]
```diff
@@ -116,12 +117,19 @@ func parseEncryptedFull(input string) (*JsonWebEncryption, error) {
 		return nil, err
 	}
 
-	obj := &JsonWebEncryption{}
-	obj.original = &parsed
-	obj.unprotected = parsed.Unprotected
+	return parsed.sanitized()
+}
+
+func (parsed *rawJsonWebEncryption) sanitized() (*JsonWebEncryption, error) {
```
Comment (sul3n3t) [inline jwe.go]: It is private, but I think a method-level comment might be helpful.

File: jwk_test.go [hunk 3]
```diff
@@ -17,40 +17,240 @@
...
+		},
+		nil,
+	}
+
+	for _, key := range keys {
+		jwk := JsonWebKey{Key: key}
+		_, err := jwk.MarshalJSON()
 		if err == nil {
-			t.Error("parser incorrectly parsed invalid key", jwk)
+			t.Error("managed to serialize invalid key")
```
Comment (sul3n3t) [inline jwk_test.go]: For all the looped test cases, it would be helpful to output which case failed with the error message.

**Comment by sul3n3t** [comment]
There's a lot here. I don't have too many comments. Aside from those I already added, consider changing the signatures for methods like `fromRsaPublicKey` to look like constructors and not method receivers. Comments for the sanitized methods would be helpful.

Looks pretty good. :+1:

**Pull request merged**